drivers/i2c/i2c_rpi_pico_pio.c

```diff
@@ -0,0 +1,602 @@
+/*
+ * Copyright (c) 2024 Yiding Jia <yiding.jia@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Portions Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/pinctrl.h>
+
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+
+#include <hardware/claim.h>
+
+LOG_MODULE_REGISTER(i2c_pio);
+
+#define DT_DRV_COMPAT raspberrypi_pico_pio_i2c
+
+struct i2c_pio_config {
+	const struct device *piodev;
+	uint sm;
+	const struct pinctrl_dev_config *pin_cfg;
+	uint32_t bitrate;
+};
+
+struct i2c_pio_data {
+	struct k_mutex lock;
+	uint32_t config;
+};
+
+/**
+ * @name I2C PIO Program
+ * Based on pico example code:
+ * https://github.com/raspberrypi/pico-examples/blob/master/pio/i2c/i2c.pio
```

Comment:
this program is actually BSD-3 licensed -- you can't just take it and implicitly relicense it as Apache 2. This probably needs some kind of TSC approval, or maybe PIO based driver should be in a dedicated module or something?

Reply:
Hm. Originally had the original license in this file but the linters didn't like that at all, so I meant to ask what should be done in this situation, but forgot about it. The C code interfacing with the PIO code is also based on the corresponding C code in the example as well, so if we want to be really diligent about it that should also be licensed BSD3. 

Is it not possible to have BSD3 code in tree? I thought BSD3 is compatible with Apache2, as long as we include the original license somewhere.

Reply:
Since the BSD-3 source code already exists, it seems like there should be no problem as long as the license is stated correctly.

https://github.com/zephyrproject-rtos/zephyr/blob/c544e4d08b43d0bcbfa22e380a899d804279fd6a/drivers/sensor/bosch/bmi08x/bmi08x_config_file.h#L5

Reply:
@yiding 

Could you put just the BSD-3 part into a separate file? (I think it would be fine to just make it a local header file.

The BSD-3 code already exists, so I think it would be best to explain it to TSC and include it as BSD-3.

Reply:
I don't think it would appease the license checker any more than this would.

I'll ask in this discord what's the best way forward here is. The docs are not clear.

Reply:
Understood. I'll follow their policy.

Reply:
"Importing code into the Zephyr OS from other projects that use a license other than the Apache 2.0 license needs to be fully understood in context and approved by the Zephyr governing board.":

https://docs.zephyrproject.org/latest/contribute/guidelines.html#licensing